File: backends/spring-java/src/main/java/dev/stackverse/backend/account/UserAccountFilter.java
package dev.stackverse.backend.account;

import dev.stackverse.backend.common.Logging;
import dev.stackverse.backend.message.MessageLocalizer;
import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.util.Map;
import org.slf4j.LoggerFactory;
import org.slf4j.event.Level;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken;
import org.springframework.web.filter.OncePerRequestFilter;
import tools.jackson.databind.ObjectMapper;

/**
 * Runs after JWT auth: upserts the caller account and rejects blocked accounts.
 */
public class UserAccountFilter extends OncePerRequestFilter {
    private final UserAccountService accountService;
    private final MessageLocalizer localizer;
    private final ObjectMapper objectMapper;
    private final org.slf4j.Logger log = LoggerFactory.getLogger(getClass());

    public UserAccountFilter(UserAccountService accountService, MessageLocalizer localizer, ObjectMapper objectMapper) {
        this.accountService = accountService;
        this.localizer = localizer;
        this.objectMapper = objectMapper;
    }

    @Override
    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain)
        throws ServletException, IOException {
        var authentication = SecurityContextHolder.getContext().getAuthentication();
        if (authentication instanceof JwtAuthenticationToken token) {
            UserAccount account = accountService.recordSeen(token.getName());
            if (account.getStatus() == UserAccountStatus.BLOCKED) {
                Logging.logEvent(
                    log,
                    Level.WARN,
                    "blocked_user_rejected",
                    "denied",
                    "Refused a request from a blocked account",
                    "actor", token.getName()
                );
                response.setStatus(HttpStatus.FORBIDDEN.value());
                response.setContentType(MediaType.APPLICATION_PROBLEM_JSON_VALUE);
                objectMapper.writeValue(
                    response.getOutputStream(),
                    Map.of(
                        "type", "about:blank",
                        "title", "Forbidden",
                        "status", HttpStatus.FORBIDDEN.value(),
                        "detail", localizer.localize("error.account.blocked", request)
                    )
                );
                return;
            }
        }
        filterChain.doFilter(request, response);
    }
}

File: backends/spring-kotlin/src/main/kotlin/dev/stackverse/backend/account/UserAccountFilter.kt
package dev.stackverse.backend.account

import dev.stackverse.backend.common.logEvent
import dev.stackverse.backend.message.MessageLocalizer
import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.slf4j.LoggerFactory
import org.slf4j.event.Level
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken
import org.springframework.web.filter.OncePerRequestFilter
import tools.jackson.databind.ObjectMapper

/**
 * Runs right after JWT authentication: upserts the caller's account row (SPEC rule 16)
 * and rejects blocked accounts with a localized 403 problem document (rule 17).
 * Registered inside the security filter chain — deliberately not a `@Component`,
 * which would make Boot register it a second time as a plain servlet filter.
 */
class UserAccountFilter(
    private val accountService: UserAccountService,
    private val localizer: MessageLocalizer,
    private val objectMapper: ObjectMapper,
) : OncePerRequestFilter() {

    // `log`, not `logger` — the inherited GenericFilterBean field keeps that name
    private val log = LoggerFactory.getLogger(javaClass)

    override fun doFilterInternal(request: HttpServletRequest, response: HttpServletResponse, filterChain: FilterChain) {
        val authentication = SecurityContextHolder.getContext().authentication
        if (authentication is JwtAuthenticationToken) {
            val account = accountService.recordSeen(authentication.name)
            if (account.status == UserAccountStatus.BLOCKED) {
                log.logEvent(
                    Level.WARN, "blocked_user_rejected", "denied", "Refused a request from a blocked account",
                    "actor" to authentication.name,
                )
                response.status = HttpStatus.FORBIDDEN.value()
                response.contentType = MediaType.APPLICATION_PROBLEM_JSON_VALUE
                objectMapper.writeValue(
                    response.outputStream,
                    mapOf(
                        "type" to "about:blank",
                        "title" to "Forbidden",
                        "status" to HttpStatus.FORBIDDEN.value(),
                        "detail" to localizer.localize("error.account.blocked", request),
                    ),
                )
                return
            }
        }
        filterChain.doFilter(request, response)
    }
}
